lib/http.c

```diff
@@ -1654,6 +1654,15 @@ CURLcode Curl_http_done(struct Curl_easy *data,
     return CURLE_GOT_NOTHING;
   }
 
+  if(status == CURLE_OK && data->req.size == -1 && !data->req.ignore_cl &&
+     (
+#ifdef USE_NGHTTP2
+      (conn->httpversion == 20 && http->close_handled) ||
+#endif
```

Comment:
This *probably* needs more attention to also work with hyper (which supports h2 without nghttp2) and HTTP/3, but I can't tell exactly how off the top of my head.

Reply:
I've somewhat improved on the logic and added to doc for CURLINFO_CONTENT_LENGTH_DOWNLOAD{,_T}. I don't know how it affects hyper. Basically I think total download size should only be set when the length can actually be determined. Notably I tried to exclude the read-until-close where there is no termination point other than close. One could argue that total download size should be set in all cases of successful transfer but having close as the only termination point seems iffy to me, for all we know it could be truncated.

Reply:
Your test 1467 works with hyper too, so it looks good!

Reply:
> Your test 1467 works with hyper too, so it looks good!

That test is only for chunked encoding, so it does not cover the case of HTTP/2 and Hyper. I have some logic for nghttp2 if curl is handling http/2 but not hyper. I just checked and unfortunately the issue is still present when hyper the http backend:

~~~
owner@ubuntu1604-x64-vm:~/curl/tests$ ../src/curl -o /dev/null -# https://pastebin.com/raw/HaAPSSR2
##O=#  #                                                                       
owner@ubuntu1604-x64-vm:~/curl/tests$ ../src/curl -V
curl 7.82.0-DEV (x86_64-pc-linux-gnu) libcurl/7.82.0-DEV OpenSSL/1.1.1i zlib/1.2.8 nghttp2/1.43.0 librtmp/2.3 Hyper/0.14.13 OpenLDAP/2.4.42
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS Debug HSTS HTTP2 HTTPS-proxy IPv6 Largefile libz NTLM NTLM_WB SSL TLS-SRP TrackMemory UnixSockets
~~~

There needs to be some way to determine that the HTTP/2 transfer from hyper was fully downloaded (end of stream reached). I made it check http->close_handled for nghttp2 so perhaps hyper needs something similar.